### PR TITLE
SmallRye Context Propagation 1.0.16

### PR DESCRIPTION
**Pull request by FroMage**

### CODE REVIEW

**Comment by FroMage** [comment]
Fixes #11378 as well.

**Review by manovotn — APPROVED** [review]

**Comment by FroMage** [comment]
@manovotn turns out the CI for quarkus-arc-deployment fails with:

```java
java.lang.NoSuchMethodError: io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest$SynthBean.<init>(Ljava/lang/String;)V
	at io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_Bean.create(SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_Bean.zig:139)
	at io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_Bean.create(SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_Bean.zig:154)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext.access$000(AbstractSharedContext.java:14)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_ClientProxy.arc$delegate(SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_ClientProxy.zig:110)
	at io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_ClientProxy.getString(SynthProxiableBeanWithoutNoArgConstructorTest_SynthBean_d2489e99512ed6afb3250b515819124dd8be8184_Synthetic_ClientProxy.zig:229)
	at io.quarkus.arc.test.unproxyable.SynthProxiableBeanWithoutNoArgConstructorTest.testSyntheticBean(SynthProxiableBeanWithoutNoArgConstructorTest.java:70)

```

**Comment by FroMage** [comment]
It also fails on my machine locally.

**Comment by manovotn** [comment]
@FroMage my bad, didn't push my local changes - the test bean class needs to be static.

File: extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/SynthProxiableBeanWithoutNoArgConstructorTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    public void testSyntheticBean() {
+        InstanceHandle<SynthBean> instance = Arc.container().instance(SynthBean.class);
+        assertTrue(instance.isAvailable());
+        assertEquals("foo", instance.get().getString());
+    }
+
+    @Vetoed
+    public class SynthBean {
```
Comment (manovotn) [inline extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/SynthProxiableBeanWithoutNoArgConstructorTest.java:74]:
```suggestion
    static class SynthBean {
```

Comment (manovotn) [reply extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/SynthProxiableBeanWithoutNoArgConstructorTest.java:74]: @FroMage this should do it. Sorry, it's my bad, I forgot to push that into my branch on Thu once I figured why it fails...

**Comment by FroMage** [comment]
Haha, great, thanks, no worries :)

**Comment by FroMage** [comment]
> 2020-08-17T14:12:30.2312122Z java.lang.LinkageError: loader constraint violation: when resolving method 'void io.smallrye.context.Jdk9CompletableFutureWrapper.<init>(io.smallrye.context.SmallRyeThreadContext, java.util.concurrent.CompletableFuture, java.util.concurrent.Executor, boolean)' the class loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @76ac879c of the current class, io/smallrye/context/impl/JdkSpecificImpl, and the class loader 'app' for the method's defining class, io/smallrye/context/Jdk9CompletableFutureWrapper, have different Class objects for the type io/smallrye/context/SmallRyeThreadContext used in the signature (io.smallrye.context.impl.JdkSpecificImpl is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @76ac879c, parent loader 'platform'; io.smallrye.context.Jdk9CompletableFutureWrapper is in unnamed module of loader 'app')

DAMNIT

**Comment by FroMage** [comment]
@dmlloyd or @stuartwdouglas do you happen to know if there's anything obvious I should be looking at for errors like these? This is due to a newly MultiRelease jar with JDK9-specific classes.

**Comment by stuartwdouglas** [comment]
I will look into it.

**Comment by stuartwdouglas** [comment]
I have pushed a fix to this PR, lets see what CI says.

**Comment by FroMage** [comment]
Thanks. Now failures in devtools :(

**Comment by FroMage** [comment]
This works on my machine so let's restart CI.

**Comment by stuartwdouglas** [comment]
I pushed a fix, there were some legitimate failures

**Comment by FroMage** [comment]
Now it's the MP-CP TCK failing, which is a problem because we've added new methods in SR-CP that will end up in the spec, but we can't release the spec ATM, so we can't release a new version of the TCK which fixes those tests.

**Comment by FroMage** [comment]
So one of the TCK failure is a real issue and I'll wait for https://github.com/smallrye/smallrye-context-propagation/pull/195 to be merged so I can pull the new release.

The other two I should ignore until we can release a new TCK.

@kenfinnigan do you happen to know how I can ignore specific test methods from the TCK? I tried the obvious surefire config, but got:

```
Method filter prohibited in includes|excludes|includesFile|excludesFile parameter: ThreadContextTest#withContextCaptureDependentCompletableFuturesRunWithContext
```

**Comment by gsmet** [comment]
@FroMage this is done like that in other MP TCKs: https://github.com/quarkusio/quarkus/blob/master/tcks/microprofile-rest-client/pom.xml#L34

**Comment by FroMage** [comment]
This excludes an entire file, I can't do that I only need to exclude two tests in that class.

**Comment by kenfinnigan** [comment]
I think the only way to exclude test methods within a class is to create a suite.xml with more specific include/exclude

**Comment by manovotn** [comment]
@FroMage I think you are going to need something like this - https://github.com/eclipse-ee4j/cdi-tck/blob/master/impl/src/main/resources/tck-tests.xml
Which you then reference from the pom this way - https://github.com/weld/core/blob/master/jboss-tck-runner/pom.xml#L240-L248

**Comment by FroMage** [comment]
Thanks @manovotn it seems to work, except now I seem to have test methods run in parallel, which the Quarkus Arquillian extension doesn't support so it fails at weird places.

**Comment by manovotn** [comment]
Hmm, can't that be disabled? In their docs i see several options but nothing like 'none'. Or maybe it's fork count rather than parallel?

**Comment by FroMage** [comment]
I'm not sure. I guess I could confirm it's due to threading first. Certainly the logs indicate something very fishy, but I'm not sure how this could be triggered by a `test.xml` file.

**Comment by FroMage** [comment]
OK, we're on for 1.0.16 now, and I've found a way to disable parallel test execution, but at the cost of listing all test classes. Since this is only temporary until we can release a new MP-CP, I think it's acceptable.

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java [hunk 2]
```diff
@@ -195,6 +195,7 @@ public void execute(BuildContext context) {
             Thread.currentThread().setContextClassLoader(runningQuarkusApplication.getClassLoader());
             // Instantiate the real test instance
             testInstance = TestInstantiator.instantiateTest(testJavaClass, runningQuarkusApplication.getClassLoader());
+            System.err.println("Setting testInstance to " + testInstance);
```
Comment (manovotn) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java:198]: @FroMage forgot to erase debugging evidence? ;-)

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java [hunk 3]
```diff
@@ -259,6 +260,7 @@ public void undeploy(Archive<?> archive) throws DeploymentException {
                 Thread.currentThread().setContextClassLoader(runningApp.get().getClassLoader());
             }
             testInstance = null;
+            System.err.println("UnSetting testInstance");
```
Comment (manovotn) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java:263]: Same as above

**Review by manovotn — CHANGES_REQUESTED** [review]

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java (hunk 2 above)
Comment (FroMage) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java:198]: DAMNIT

**Comment by manovotn** [comment]
> I've found a way to disable parallel test execution, but at the cost of listing all test classes.

I agree that it's acceptable as temporary solution (maybe add a comment about remove that once we update version?). But it's really puzzling how this causes/prevents parallel execution :D

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java (hunk 2 above)
Comment (FroMage) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java]: OK should be fixed, thanks a lot :)

**Comment by FroMage** [comment]
> I agree that it's acceptable as temporary solution (maybe add a comment about remove that once we update version?). But it's really puzzling how this causes/prevents parallel execution :D

StackOverflow is full of people noticing this issue and trying to get rid of it with no luck. They said surefire even goes as far as generating one test suite per class to make sure testng doesn't do that, which specifying a `test.xml` file probably defeats.

**Comment by FroMage** [comment]
Last failures are due to me forgetting the `test.xml` file, and @stuartwdouglas being optimistic that `java.version` starts with `8.` when before JDK9 they started with `1.8` ;) I'll fix this.

**Comment by gsmet** [comment]
Rebased and cleaned up the commits a bit to reorder them to avoid failures.

@manovotn are you OK with this PR now?

**Review by manovotn — APPROVED** [review]

**Comment by manovotn** [comment]
I thought I green-stamped this long ago, but it was probably lost with the new changes coming in, sorry for the delay.

**Comment by gsmet** [comment]
Rebased to fix a conflict and get the latest CI fixes.

**Comment by gsmet** [comment]
The MP tests took more than 2 hours whereas they are usually done in 40 minutes.

Restarting a build to see if it's only a glitch or if there is an issue.

**Comment by gsmet** [comment]
@FroMage the MP TCK tests are consistently taking 2 hours+ which is far worse than the usual runs (I have checked a few runs and it was more like 40 minutes). Could you have a look at what's going on? I don't feel like merging before this is diagnosed.

**Comment by FroMage** [comment]
> @FroMage the MP TCK tests are consistently taking 2 hours+ which is far worse than the usual runs (I have checked a few runs and it was more like 40 minutes). Could you have a look at what's going on? I don't feel like merging before this is diagnosed.

OK.

**Comment by FroMage** [comment]
Runs in 17s here :(

**Comment by FroMage** [comment]
Actually, that was the MP-CP TCK, let's run all the TCKs perhaps it happens in other TCKs.

**Comment by FroMage** [comment]
Ah, OK, looks like `org.eclipse.microprofile.fault.tolerance.tck.AsyncCancellationTest` hangs, even locally.

**Comment by FroMage** [comment]
So, all tests pass and complete, but there's a daemon thread remaining with this stack:

```
Daemon Thread [FutureExecution{executor-thread-2}] (Suspended)	
	Unsafe.park(boolean, long) line: not available [native method]	
	LockSupport.park(Object) line: 175	
	ThreadPoolBulkhead$BulkheadTask(FutureTask<V>).awaitDone(boolean, long) line: 429	
	ThreadPoolBulkhead$BulkheadTask(FutureTask<V>).get() line: 191	
	ThreadPoolBulkhead<V>.apply(InvocationContext<Future<V>>) line: 63	
	ThreadPoolBulkhead<V>.apply(InvocationContext) line: 27	
	GeneralMetricsRecorder<V>.apply(InvocationContext<V>) line: 17	
	FutureExecution<V>.lambda$apply$0(InvocationContext) line: 27	
	2014777783.call() line: not available [local variables unavailable]	
	NamedFutureTask<T>(FutureTask<V>).run() line: 266 [local variables unavailable]	
	NamedFutureTask<T>.run() line: 20	
	SmallRyeThreadContext$ContextualRunnable.run() line: 58	
	ContextClassLoaderSavingRunnable.run() line: 35	
	ViewExecutor$TaskWrapper.run() line: 320	
	ContextClassLoaderSavingRunnable.run() line: 35	
	EnhancedQueueExecutor.safeRun(Runnable) line: 2046	
	EnhancedQueueExecutor$ThreadBody.doRunTask(Runnable) line: 1578	
	EnhancedQueueExecutor$ThreadBody.run() line: 1452	
	ThreadLocalResettingRunnable(DelegatingRunnable).run() line: 29	
	ThreadLocalResettingRunnable.run() line: 29	
	JBossThread(Thread).run() line: 748	
	JBossThread.run() line: 479	
```

**Comment by FroMage** [comment]
And indeed, this passes with the previous SR-CP release.

**Comment by FroMage** [comment]
OK found it.

@Ladicek or @michalszynkiewicz  perhaps you can make sense of it?

What happens is that on Quarkus we forward all async tasks by default to the Quarkus Executor, so if you inject `@ManagedExecutor` it will not create a new executor but forward to it. When we build the `SmallRyeContextManager` we set the default executor to the Quarkus one.

Now, in `ContextPropagationExecutorFactory` in SR-FT you use this:

```java
    @Override
    public ExecutorService createCoreExecutor(int size) {
        return ManagedExecutor.builder().maxAsync(size).build();
    }

    @Override
    public ExecutorService createExecutor(int coreSize, int size) {
        SmallRyeManagedExecutor.Builder builder = (SmallRyeManagedExecutor.Builder) ManagedExecutor.builder();
        return builder.maxAsync(size).build();
    }
```

Which call the builder, but don't set any default executor on it with `withExecutorService`, so previously, there was a bug where it would create a new executor service instead of using the default one set on the `SmallRyeContextManager`.

And your test would not hang. Now that I properly set the default executor to the quarkus one, your test hangs for some weird reason (I mean, the test passes, but a deamon thread is left waiting).

Perhaps this is due to your test having a `maxAsync` of `1000` which I'm pretty sure the Quarkus Executor won't support OOTB, after all we're creating a "view" on the executor, so we'll only be able to scale "up to" what the target executor supports, and that's set by the Quarkus config. So there's a mismatch here.

Now, the MP-CP spec doesn't really specify default executors yet because this is an SR addition for now, though we're moving it up to the spec, but we can't update the spec.

I could revert to the previous behaviour, but I think it's wrong. You guys could alternately call `withExecutorService(null)` to remove the default Quarkus Executor and get a new thread pool.

Not sure how to proceed here. Ping @dmlloyd who gave me the executor view and may have an opinion.

**Comment by Ladicek** [comment]
SmallRye Fault Tolerance expects completely new executors that it can assume control over :-)

If `ManagedExecutor.builder().build()` doesn't create a new thread pool but reuses some global one, that itself is a strange behavior to me. On the other hand, I see that MicroProfile Context Propagation allows it (though the language is not exactly clear), so it was a mistake on SRye FT side.

I could use the SmallRye ConProp extension that you mention (`withExecutorService(null)`), no problem with that, but are you sure that end users won't be surprised by this change?

**Comment by dmlloyd** [comment]
> SmallRye Fault Tolerance expects completely new executors that it can assume control over :-)

In what regard?  What contract is being expected of the executor, beyond accepting tasks for deferred execution?

**Comment by Ladicek** [comment]
Size, mainly, and lifecycle.

We need to be able to get a thread pool of given size, for two reasons: for thread pool style bulkheads (which I'd eventually like to reimplement, because it's a waste of resources to create a thread pool for each bulkhead, but that's how it is currently), and for the main async execution facility (where we already expose a configuration option to set a size; that could eventually go away as well, but that most likely needs to be an integration concern that can't be solved in SRye FT itself).

Secondarily, given we need to create our own thread pools because of the above, we also need to be able to shut them down. (We might also want this to let thread pool style bulkheads shut down their thread pools if they are not used, but that's speculative and I really don't intend to do anything like that.)

**Comment by FroMage** [comment]
I think it's useful that on Quarkus we prevent people from creating their own executors by having a default executor.

It's unfortunate that it now implies that if you really must create new thread pools you'll have to call `.withExecutorService(null)`, as in your case.

As for lifecycle, both the view on the default Quarkus executor or a new executor support `shutdown` (in the first case, without breaking the default Quarkus executor) so as far as you're concerned you can control their lifecycle the same.

I really wonder if you'll ever really need a separate thread pool than the Quarkus one though. Need as in: it can't work if we put all our async threads in the Quarkus pool, it needs to be bigger/separate. That's a question I can't really answer.

**Comment by Ladicek** [comment]
So technically, I can go and create my own thread pool and contextualize it using `ThreadContext` and be done with it, and noone would ever know :-) Everyone else can do the same, too. Attempting to prevent users from creating their own thread pools is futile, in my opinion -- you'll just end up with people frustrated that they need to write more code to achieve the same thing. Also, in my naive thinking, `ManagedExecutor` lets me build an executor = it lets me build a thread pool. Personally, I kinda equate those two terms. The fact that I can set the size reinforces this.

That said -- to reproduce the issue, it is enough to check out this PR, build it and run the MP FT TCK with it?

One other thing we can do -- provide Quarkus-specific `ExecutorFactory`. What's inside SRye FT is a hopefully-sensible default that is meant to be overridden by an integrator.

**Comment by FroMage** [comment]
> So technically, I can go and create my own thread pool and contextualize it using ThreadContext and be done with it, and noone would ever know :-) Everyone else can do the same, too. Attempting to prevent users from creating their own thread pools is futile, in my opinion -- you'll just end up with people frustrated that they need to write more code to achieve the same thing. Also, in my naive thinking, ManagedExecutor lets me build an executor = it lets me build a thread pool. Personally, I kinda equate those two terms. The fact that I can set the size reinforces this.

Yes, our goal was not to make it impossible, but we wanted to avoid that people would create thread pools left and right just to get context propagation via `ManagedExecutor`. I never liked `ManagedExecutor` and the fact that it was the go-to place to get `CompletionStage` and `CompletableFuture` which forced us to create thread pools even if we never needed the `Async` methods, like in most reactive programming. Using the Quarkus Executor mitigated that.

It could seem enough that the default injected `ManagedExecutor` would use the Quarkus executor, but the minute anyone wanted a different set of contexts to propagate, they would need to create their own `ManagedExecutor` and would risk getting a new thread pool to manage without necessarily requiring it.

Now that we can contextualise CF/CS which would still use the Quarkus executor in the spec (a new change) then perhaps it makes sense to add `newIncompleteFuture`/`failedStage`/`failedFuture`/`completedStage` on `ThreadContext` to mirror their existing `ManagedExecutor` locations and we could tell people to never build a `ManagedExecutor` unless they want a new executor. This way we could tell people to stick to `ThreadContext` unless they need a new thread pool (or to access the Quarkus thread pool for the injected version).

But we can't release a new MP-CP spec ATM, so those new methods would have to live on `SmallRyeThreadContext` for a while.

> That said -- to reproduce the issue, it is enough to check out this PR, build it and run the MP FT TCK with it?

Yes

> One other thing we can do -- provide Quarkus-specific ExecutorFactory. What's inside SRye FT is a hopefully-sensible default that is meant to be overridden by an integrator.

You mean that it may be doable to reuse the Quarkus executor for FT on Quarkus instead of creating your own?

**Comment by Ladicek** [comment]
OK, I understand. I really like your idea of "tell people to never build a `ManagedExecutor` unless they want a new executor. This way we could tell people to stick to `ThreadContext` unless they need a new thread pool", and it's too bad we can't do this.

> You mean that it may be doable to reuse the Quarkus executor for FT on Quarkus instead of creating your own?

I'm saying Quarkus can override the SRye FT's `ContextPropagationExecutorFactory` and do whatever is best. I'm still not sure what's best :-)

**Comment by Ladicek** [comment]
FYI, I can confirm that with this change https://github.com/smallrye/smallrye-fault-tolerance/pull/277, the MP FT TCK in Quarkus passes on top of this PR.

I can do a release of SRye FT with this, if you want. (Tomorrow! :-) )

**Comment by gsmet** [comment]
If we want it in CR1, I would need to have everything merged before noon Paris time. Meaning if it's possible to push the SRFT upgrade directly to this PR as early as possible, that would be the quickest way to get CI to run on both of the changes and have things merged on time.

**Comment by gsmet** [comment]
Also better rebase to latest master, given the number of CI issues we fixed.

**Comment by stuartwdouglas** [comment]
We definitely do not want extensions creating there own thread pools. More threads == more memory, which is why Quarkus attempts to use a shared executor whenever possible.

I don't think you need to use a new thread pool for bulkheads, you can just submit the tasks to the shared thread pool but put a limit on the number of currently executing requests.

**Comment by Ladicek** [comment]
I know we can implement thread pool style bulkheads on top of a shared thread pool. We don't have that implementation yet. (Well, we partially do. We have 2 implementations of thread pool style bulkheads, one uses its own thread pools, which is inefficient by definition, the other using a shared thread pool and emulating a queue by submitting all tasks to the thread pool and letting those that should be queued wait on a semaphore, which is also terribly inefficient.)

**Comment by FroMage** [comment]
@gsmet let's not rush this issue, it can wait for the next release, I'd rather get this right.

**Comment by FroMage** [comment]
OK, so how about we keep the new default of letting the platform define whether you get a new thread pool by default, and let you request a new executor with `SmallRyeManagedExecutor.Buidler.withExecutorService(null)` and a new alias `SmallRyeManagedExecutor.Buidler.withNewExecutorService()` to make it clearer?

And I'll add methods to deal with new CS/CF from `ThreadContext` so we can avoid `ManagedExecutor` entirely unless you must deal with executors.

**Comment by Ladicek** [comment]
That makes sense to me. If you could later push those improvements to MP ConProp, that would be great as well :-)

**Comment by Ladicek** [comment]
Additionally, I'll look into the thread pool style bulkheads today, it might be a quick win.

**Comment by gsmet** [comment]
> let's not rush this issue, it can wait for the next release, I'd rather get this right.

Yeah totally agreed, I removed the 1.8 milestone a bit before your message :).

**Comment by FroMage** [comment]
> That makes sense to me. If you could later push those improvements to MP ConProp, that would be great as well :-)

On it at https://github.com/smallrye/smallrye-context-propagation/pull/202

**Comment by gsmet** [comment]
Any progress on this?

**Comment by Ladicek** [comment]
I made a bit of a progress on SmallRye Fault Tolerance side, here: https://github.com/smallrye/smallrye-fault-tolerance/pull/279 With that PR, SRye FT still creates 1 thread pool, but only 1 and uses it for everything. I need to come up with an API how to get that thread pool from the integrator.

**Comment by Ladicek** [comment]
I'm just wondering, is the worker thread pool (the one that Quarkus should provide to SRye Fault Tolerance for executing async actions etc.) available for injection in the ArC container? @FroMage @mkouba do you know?

(I'm thinking the integration API should probably be CDI based. Similarly to `FallbackHandlerProvider`, @mkouba you might remember that one :-) )

**Comment by Ladicek** [comment]
I've just added one more commit to https://github.com/smallrye/smallrye-fault-tolerance/pull/279 to overhaul the thread pool integration point. This should allow Quarkus to provide its worker thread pool to SmallRye Fault Tolerance, which will then be used for all async tasks.

**Comment by mkouba** [comment]
> I'm just wondering, is the worker thread pool available for injection in the ArC container? 

@Ladicek `io.smallrye.faulttolerance.ExecutorFactory` was a service provider so no injection.

> I'm thinking the integration API should probably be CDI based.

Well, I don't have a strong opinion. A service provider with a priority should work too but CDI is probably more convenient.

**Comment by Ladicek** [comment]
> > I'm just wondering, is the worker thread pool available for injection in the ArC container?
> 
> @Ladicek `io.smallrye.faulttolerance.ExecutorFactory` was a service provider so no injection.

What I currently have is CDI based, not a service loader anymore. But before I fully commit to it, I wonder if that makes sense. That is, if the worker thread pool available for injection :-)

**Comment by mkouba** [comment]
> is the worker thread pool available for injection? 

@Ladicek Do you mean the Vertx worker thread pool?

**Comment by Ladicek** [comment]
I thought the entire idea is that there's exactly one worker thread pool in all of Quarkus, but I'm no expert :-) I mean: yes, I [probably] mean the Vert.x worker thread pool.

**Comment by mkouba** [comment]
> I thought the entire idea is that there's exactly one worker thread pool in all of Quarkus, but I'm no expert :-) I mean: yes, I [probably] mean the Vert.x worker thread pool.

I think that there are currently at least two "main" thread pools for blocking tasks. The Vertx workers and the `ExecutorService` provided by `io.quarkus.deployment.builditem.ExecutorBuildItem`. The latter one is used for example to execute `@Scheduled` methods by default, it's wrapped by the SM CP `org.eclipse.microprofile.context.ManagedExecutor` or used for CDI async events. There are of course others, for example Quartz has its own thread pool.

AFAIK none of these two is directly injectable. But it would be easy to make the `ExecutorService` from the `ExecutorBuildItem` injectable, if needed.

**Comment by Ladicek** [comment]
I totally need the one that's backing the ConProp's `ManagedExecutor`. FYI here's the Fault Tolerance refactoring: https://github.com/smallrye/smallrye-fault-tolerance/pull/279

**Comment by Ladicek** [comment]
@manovotn helped me realize that I can just inject MP ConProp's `ManagedExecutor`, which is exactly what Fault Tolerance needs, so we should be fine here. Well, when SRye FT 5.0.0 is released, that is :-)

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
The `Quarkus CI / MicroProfile TCKs Tests` failed but I don't think it's related to these changes?

**Comment by Ladicek** [comment]
If I remember correctly, it is quite related :-) I have a preliminary PR to SmallRye Fault Tolerance that makes everything run on a single thread pool provided by an integrator, but I've uncovered some issues with how liberally we submit tasks to thread pools and wait for their completion. I'm working on a fix for that.

**Comment by Ladicek** [comment]
OK, so I have reimplemented and optimized bulkheads in SmallRye Fault Tolerance (optimized as in: bulkheads no longer submit any tasks to an executor). It's all built on top of the changes I did to implement MicroProfile Fault Tolerance 3.0, so wouldn't be exactly easy to backport to current `master`. But MP FT 3.0 should be released "soon" so I hope that's fine.

**Comment by FroMage** [comment]
I'll make a new SR-CP release and rebase this branch ASAP.

**Comment by FroMage** [comment]
Upgraded to latest SR-CP version, rebased on master, fixed conflicts. I have to write some docs about the new features.

**Comment by FroMage** [comment]
Failed with:

```
2020-10-21T14:34:50.1749165Z 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#registerBeans threw an exception: javax.enterprise.inject.spi.DefinitionException: Interceptor has no bindings: io.smallrye.context.impl.cdi.SmallRyeCurrentThreadContextInterceptor
2020-10-21T14:34:50.1762017Z 	at io.quarkus.arc.processor.Interceptors.createInterceptor(Interceptors.java:44)
2020-10-21T14:34:50.1766515Z 	at io.quarkus.arc.processor.BeanDeployment.findInterceptors(BeanDeployment.java:1052)
2020-10-21T14:34:50.1776096Z 	at io.quarkus.arc.processor.BeanDeployment.registerBeans(BeanDeployment.java:220)
2020-10-21T14:34:50.1786451Z 	at io.quarkus.arc.processor.BeanProcessor.registerBeans(BeanProcessor.java:115)
2020-10-21T14:34:50.1795862Z 	at io.quarkus.arc.deployment.ArcProcessor.registerBeans(ArcProcessor.java:381)
2020-10-21T14:34:50.1803462Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-10-21T14:34:50.1810664Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-10-21T14:34:50.1874797Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-10-21T14:34:50.1877039Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2020-10-21T14:34:50.1878316Z 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:936)
2020-10-21T14:34:50.1879323Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
2020-10-21T14:34:50.1881055Z 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
2020-10-21T14:34:50.1882920Z 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
2020-10-21T14:34:50.1884456Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
2020-10-21T14:34:50.1885895Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
2020-10-21T14:34:50.1886845Z 	at java.lang.Thread.run(Thread.java:748)
2020-10-21T14:34:50.1887518Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

So I must have forgotten something.

**Comment by manovotn** [comment]
As per out Zulip conversation, the annotation is not indexed. This is likely due to missing `beans.xml` in the `api` archive of SR CP.

**Comment by FroMage** [comment]
Should be fixed by a new release. Rebased again.

**Comment by FroMage** [comment]
Apparently the FT TCK got killed. Not sure if it hanged or CI just ran out of patience.

**Comment by gsmet** [comment]
@FroMage it usually only takes ~40 minutes so there's something fishy if it got cancelled after 2 hours.

Looks like the same issue we had previously.

**Comment by FroMage** [comment]
@gsmet thanks a lot, you're right.

So @Ladicek we need to couple this with a new SR-FT release?

**Comment by Ladicek** [comment]
@FroMage That would be best, yea. I'm not sure when MP FT 3.0 is released, and I have all the changes on my `next` branch which is for MP FT 3.0 :-/ I can try to backport the thread pool changes to `master`, if you want.

**Comment by FroMage** [comment]
I'm going to make a new SR-CP release with support for Java 12 anyway, but are we talking days or months?

**Comment by Ladicek** [comment]
I asked @radcortez and it seems it's not "days". I'll grab a bottle and do some backporting -- sorry for stalling this so long.

**Comment by Ladicek** [comment]
I've added a workaround as discussed on Zulip, and released SRye FT 4.3.2: https://repo.maven.apache.org/maven2/io/smallrye/smallrye-fault-tolerance/4.3.2/ You can bump in this PR, or I can submit a separate PR, but I guess doing it here would be faster. There's no other change required, just bumping the version number in the BOM.

**Comment by Ladicek** [comment]
Actually I couldn't even bump SRye FT to 4.3.2 in isolation, because it uses the new method from SRye ConProp. So it needs to be udpated in this PR anyway :-)

**Comment by FroMage** [comment]
OK thanks a lot. Bumped it, let's see what CI says.

**Comment by FroMage** [comment]
TCK CI passed, so that's good news.

**Comment by FroMage** [comment]
WOOHOO!

**Pull request merged**